### PR TITLE
Refactor ImagingForwardBackward class

### LINKED ISSUE #284: Refactor the forward backward classes
SOPT has several "solvers" that take the operators `f_gradient` and `g_proximal` as arguments and several definitions for those operators using different norms. We could implement the learned model as a "norm" and pass it in using the current interface with minimal modifications to the rest of the code. Note that the code on [lines 234-236 of ForwardBackward.h](https://github.com/astro-informatics/sopt/blob/development/cpp/sopt/forward_backward.h#L234-L236) is specific to the L1 norm and needs to be put in a function.

Currently the definitions of different norms have a lot of copy-pasted code that should be merged while we add the learned model. The current norms for the forward backward solver are `imaging_forward_backward.h` and `l2_forward_backward.h`. PR #256 adds another one to the primal dual solver.

Create a new class that combines the classes in `imaging_forward_backward.h`, `l2_forward_backward.h` (including some of the clean-up in #256) 

Once this is done, we can proceed with
- [x] #298
- [x] #299

### PR DESCRIPTION
**Pull request by tkoskela**
Closes #284 

Related to #298 but won't close it in this PR

Created new issue #306 to be done after this

Refactor the class `ImagingForwardBackward` so that the `g_proximal` function is provided by a separate class. The public interface to provide `g_proximal` is the abstract base class `GProximal`. An implementation of the interface for the L1 proximal is added in `L1GProximal`, which holds the code specific to the L1 proximal that was previously in `ImagingForwardBackward`.

Remaining open issues
- [x] Add the `Psi` operator to the interface. It is an operator that transforms the image to a different base. In many cases it is not needed and a unit matrix can be used by default. I've kept the `GProximal` class abstract, and left the implementation to the child class(es). The `linear_transform_identity()` helper function can be used if no transform is necessary.
- [x] Implementation of `GProximal` is now chosen and injected by whoever instantiates ImagingForwardBackward. For example, see https://github.com/astro-informatics/sopt/blob/87beb2dd2d2f7cb8fe37a9cb6f8a679e8537f959/cpp/tests/inpainting.cc#L70-L79
- [x] I've added comments that try to explain the logic
- [x] All examples and tests have been updated

### CODE REVIEW

**Comment by tkoskela** [comment]
MacOS with apple-clang fails to find OpenMP on GitHub actions. Same problem appears in the development branch so something must have broken on the GA runners. Created issue #308

File: cpp/sopt/l1_g_proximal.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  //! Calls l1 proximal operator, checking for real constraints and tight frame
+  template <class T0, class T1>
+  typename proximal::L1<Scalar>::Diagnostic l1_proximal(Eigen::MatrixBase<T0> &out, Real gamma,
+                                                        Eigen::MatrixBase<T1> const &x) const {
+    return l1_proximal_real_constraint()
+      ? call_l1_proximal(out, gamma, x.real().template cast<typename T1::Scalar>())
+      : call_l1_proximal(out, gamma, x);
+  }
+
+  //! Calls l1 proximal operator, checking for thight frame
```
Comment (SJaffa) [inline cpp/sopt/l1_g_proximal.h]:
```suggestion
  //! Calls l1 proximal operator, checking for tight frame
```

File: cpp/sopt/l1_g_proximal.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    return *this;
+  }
+
+protected:
+
+  bool tight_frame_;
+  proximal::L1<Scalar> l1_proximal_;
+
+  // Helper functions for calling l1_proximal
+  //! Calls l1 proximal operator, checking for real constraints and tight frame
```
Comment (SJaffa) [inline cpp/sopt/l1_g_proximal.h]:
```suggestion
  //! Calls l1 proximal operator, checking for real constraints
```
I think this bit only checks for the real constraint, tight frame is checked in the `call_l1_proximal`

File: cpp/sopt/l1_g_proximal.h [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  proximal::L1<Scalar> const &l1_proximal() const { return l1_proximal_; }
+  L1GProximal<SCALAR> &l1_proximal(proximal::L1<Scalar> const &arg) {
+    l1_proximal_ = arg;
+    return *this;
+  }
+
+// This macro creates get/setters that point to l1_proximal
+// In practice, we end up with a bunch of functions that make it simpler to set or get values
+// associated with the two proximal operators.
+// E.g.: `paddm.l1_proximal_itermax(100).l1_proximal_tolerance(1e-4)`.
```
Comment (SJaffa) [inline cpp/sopt/l1_g_proximal.h:84]: Great to add this explanation of the macro, makes it much clearer why we have this and how it is used.

**Review by SJaffa — CHANGES_REQUESTED** [review]
My only requests were a couple of typos in the comments, but the new structure looks great, so much easier to follow the logic and should make it much easier to adapt in the future.

**Comment by CosmoMatt** [comment]
@tkoskela from what I can tell this looks great, and if @SJaffa has also taken a read through the PR we should be fine to merge into main. The only issue is this failing test, we really should try and avoid merging failing tests into main, is there any way around this?

**Review by CosmoMatt — APPROVED** [review]
This looks fine to me :)

**Review by SJaffa — APPROVED** [review]

**Pull request merged**